File: Editor/Tools/RequestScriptCompilationTool.cs
using System;
using System.Threading.Tasks;
using McpUnity.Unity;
using UnityEngine;
using UnityEditor;
using UnityEditor.Compilation;
using Newtonsoft.Json.Linq;
using McpUnity.Utils;
using McpUnity.Services;

namespace McpUnity.Tools
{
    /// <summary>
    /// Tool for requesting Unity script compilation and monitoring its progress
    /// </summary>
    public class RequestScriptCompilationTool : McpToolBase
    {
        private TaskCompletionSource<JObject> _currentTcs;
        private bool _isCompiling = false;
        private readonly IConsoleLogsService _consoleLogsService;        public RequestScriptCompilationTool(IConsoleLogsService consoleLogsService)
        {
            Name = "request_script_compilation";
            Description = "Requests Unity to recompile all scripts and reports when compilation is complete or if errors occur";
            IsAsync = true;
            _consoleLogsService = consoleLogsService;
            
            // Subscribe to compilation events
            CompilationPipeline.compilationStarted -= OnCompilationStarted;
            CompilationPipeline.compilationStarted += OnCompilationStarted;
            
            CompilationPipeline.compilationFinished -= OnCompilationFinished;
            CompilationPipeline.compilationFinished += OnCompilationFinished;
        }
        
        /// <summary>
        /// Execute the script compilation tool asynchronously
        /// </summary>
        /// <param name="parameters">Tool parameters (currently none required)</param>
        /// <param name="tcs">TaskCompletionSource to set the result</param>
        public override void ExecuteAsync(JObject parameters, TaskCompletionSource<JObject> tcs)
        {
            try
            {
                // Check if compilation is already in progress
                if (_isCompiling)
                {
                    tcs.SetResult(new JObject
                    {
                        ["success"] = false,
                        ["type"] = "text",
                        ["message"] = "Script compilation is already in progress. Please wait for it to complete before requesting another compilation.",
                        ["isCompiling"] = true
                    });
                    return;
                }

                // Check if compilation is required
                if (!EditorApplication.isCompiling)
                {
                    _currentTcs = tcs;
                    
                    McpLogger.LogInfo("Requesting script compilation...");
                    
                    // Send initial response to let the agent know compilation has started
                    var initialResponse = new JObject
                    {
                        ["success"] = true,
                        ["type"] = "text",
                        ["message"] = "Script compilation requested. Unity will now recompile all scripts. Please wait for compilation to complete before proceeding with other actions.",
                        ["status"] = "compilation_started",
                        ["note"] = "This tool will complete once compilation finishes. Monitor the Unity console for any compilation errors."
                    };
                    
                    // Request compilation
                    CompilationPipeline.RequestScriptCompilation();
                    
                    // If compilation doesn't start immediately, return a response
                    if (!EditorApplication.isCompiling)
                    {
                        // No compilation was needed
                        tcs.SetResult(new JObject
                        {
                            ["success"] = true,
                            ["type"] = "text",
                            ["message"] = "No script compilation was required. All scripts are already up to date.",
                            ["status"] = "no_compilation_needed"
                        });
                        _currentTcs = null;
                    }
                }
                else
                {
                    // Compilation is already happening
                    _currentTcs = tcs;
                    tcs.SetResult(new JObject
                    {
                        ["success"] = true,
                        ["type"] = "text",
                        ["message"] = "Script compilation is already in progress. Waiting for it to complete...",
                        ["status"] = "compilation_in_progress"
                    });
                }
            }
            catch (Exception ex)
            {
                McpLogger.LogError($"Error requesting script compilation: {ex.Message}");
                tcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                    $"Failed to request script compilation: {ex.Message}",
                    "compilation_error"
                ));
            }
        }

        /// <summary>
        /// Called when compilation starts
        /// </summary>
        /// <param name="obj">Compilation context (unused)</param>
        private void OnCompilationStarted(object obj)
        {
            _isCompiling = true;
            McpLogger.LogInfo("Script compilation started.");
        }

        /// <summary>
        /// Called when compilation finishes
        /// </summary>
        /// <param name="obj">Compilation context (unused)</param>
        private void OnCompilationFinished(object obj)
        {
            _isCompiling = false;
            
            if (_currentTcs != null)
            {
                try
                {
                    // Check for compilation errors
                    var hasErrors = HasCompilationErrors();
                    
                    if (hasErrors)
                    {
                        var errorResponse = new JObject
                        {
                            ["success"] = false,
                            ["type"] = "text",
                            ["message"] = "Script compilation completed with errors. Check the Unity Console for detailed error messages.",
                            ["status"] = "compilation_failed",
                            ["hasErrors"] = true,
                            ["note"] = "Please fix the compilation errors before proceeding. You can view detailed error information in the Unity Console window."
                        };
                        
                        _currentTcs.SetResult(errorResponse);
                        McpLogger.LogWarning("Script compilation completed with errors.");
                    }
                    else
                    {
                        var successResponse = new JObject
                        {
                            ["success"] = true,
                            ["type"] = "text",
                            ["message"] = "Script compilation completed successfully. All scripts have been recompiled without errors.",
                            ["status"] = "compilation_successful",
                            ["hasErrors"] = false
                        };
                        
                        _currentTcs.SetResult(successResponse);
                        McpLogger.LogInfo("Script compilation completed successfully.");
                    }
                }
                catch (Exception ex)
                {
                    McpLogger.LogError($"Error processing compilation completion: {ex.Message}");
                    _currentTcs.SetResult(McpUnitySocketHandler.CreateErrorResponse(
                        $"Error processing compilation result: {ex.Message}",
                        "compilation_processing_error"
                    ));
                }
                finally
                {
                    _currentTcs = null;
                }
            }
        }        /// <summary>
        /// Check if there are any compilation errors in the console
        /// </summary>
        /// <returns>True if compilation errors exist</returns>
        private bool HasCompilationErrors()
        {
            try
            {
                // Get console log entries to check for compilation errors
                var logsJson = _consoleLogsService.GetLogsAsJson("error", 0, 100, false);
                var logs = logsJson["logs"] as JArray;
                
                if (logs != null)
                {
                    foreach (var log in logs)
                    {
                        var message = log["message"]?.ToString() ?? "";
                        if (message.Contains("error CS") || 
                            message.Contains("Compilation failed") ||
                            (message.Contains("Assembly") && message.Contains("error")))
                        {
                            return true;
                        }
                    }
                }
                
                return false;
            }
            catch (Exception ex)
            {
                McpLogger.LogWarning($"Could not check for compilation errors: {ex.Message}");
                // If we can't check, assume no errors to avoid false negatives
                return false;
            }
        }

        /// <summary>
        /// Cleanup event subscriptions when the tool is disposed
        /// </summary>
        ~RequestScriptCompilationTool()
        {
            try
            {
                CompilationPipeline.compilationStarted -= OnCompilationStarted;
                CompilationPipeline.compilationFinished -= OnCompilationFinished;
            }
            catch
            {
                // Ignore exceptions during cleanup
            }
        }
    }
}

File: Server~/src/tools/requestScriptCompilationTool.ts
import * as z from 'zod';
import { Logger } from '../utils/logger.js';
import { McpUnity } from '../unity/mcpUnity.js';
import { McpServer } from '@modelcontextprotocol/sdk/server/mcp.js';
import { McpUnityError, ErrorType } from '../utils/errors.js';
import { CallToolResult } from '@modelcontextprotocol/sdk/types.js';

// Constants for the tool
const toolName = 'request_script_compilation';
const toolDescription = 'Requests Unity to recompile all scripts and reports when compilation is complete or if errors occur';
const paramsSchema = z.object({
  // No parameters required for this tool
});

/**
 * Creates and registers the Request Script Compilation tool with the MCP server
 * This tool triggers Unity script compilation and monitors its progress
 * 
 * @param server The MCP server instance to register with
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param logger The logger instance for diagnostic information
 */
export function registerRequestScriptCompilationTool(server: McpServer, mcpUnity: McpUnity, logger: Logger) {
  logger.info(`Registering tool: ${toolName}`);
    // Register this tool with the MCP server
  server.tool(
    toolName,
    toolDescription,
    paramsSchema.shape,
    async (params: any = {}) => {
      try {
        logger.info(`Executing tool: ${toolName}`, params);
        const result = await toolHandler(mcpUnity, params);
        logger.info(`Tool execution successful: ${toolName}`);
        return result;
      } catch (error) {
        logger.error(`Tool execution failed: ${toolName}`, error);
        throw error;
      }
    }
  );
}

/**
 * Handles script compilation requests
 * 
 * @param mcpUnity The McpUnity instance to communicate with Unity
 * @param params The parameters for the tool (currently unused)
 * @returns The result of the compilation request
 */
async function toolHandler(mcpUnity: McpUnity, params: any): Promise<CallToolResult> {
  try {
    // Execute the Unity tool to request script compilation
    const response = await mcpUnity.sendRequest({
      method: toolName,
      params: params || {}
    });
    
    const success = response.success || false;
    const message = response.message || 'Script compilation request completed';
    const status = response.status || 'unknown';
    const hasErrors = response.hasErrors || false;
    const isCompiling = response.isCompiling || false;
    
    // Format the response based on the compilation status
    let content = `**Unity Script Compilation**\n\n`;
    
    if (status === 'compilation_started') {
      content += `✅ **Status**: Compilation Started\n`;
      content += `📝 **Message**: ${message}\n\n`;
      content += `⚠️ **Important**: Please wait for compilation to complete before proceeding with other Unity operations.\n`;
      content += `🔍 **Monitor**: Check the Unity Console for any compilation errors or warnings.`;
    } else if (status === 'compilation_successful') {
      content += `✅ **Status**: Compilation Successful\n`;
      content += `📝 **Message**: ${message}\n\n`;
      content += `🎉 **Result**: All scripts have been successfully recompiled without errors.\n`;
      content += `✅ **Ready**: You can now proceed with other Unity operations.`;
    } else if (status === 'compilation_failed') {
      content += `❌ **Status**: Compilation Failed\n`;
      content += `📝 **Message**: ${message}\n\n`;
      content += `🐛 **Errors**: Compilation completed with errors.\n`;
      content += `🔧 **Action Required**: Please check the Unity Console and fix compilation errors before proceeding.\n`;
      content += `📋 **Tip**: Look for error messages starting with "error CS" in the Unity Console.`;
    } else if (status === 'compilation_in_progress') {
      content += `⏳ **Status**: Compilation In Progress\n`;
      content += `📝 **Message**: ${message}\n\n`;
      content += `⚠️ **Please Wait**: Compilation is already running. Wait for it to complete.`;
    } else if (status === 'no_compilation_needed') {
      content += `✅ **Status**: No Compilation Needed\n`;
      content += `📝 **Message**: ${message}\n\n`;
      content += `📄 **Info**: All scripts are already up to date.`;
    } else if (isCompiling) {
      content += `⏳ **Status**: Compilation Already Running\n`;
      content += `📝 **Message**: ${message}\n\n`;
      content += `⚠️ **Please Wait**: Another compilation is in progress.`;
    } else {
      content += `📝 **Message**: ${message}\n`;
      content += `📊 **Status**: ${status}`;
    }
    
    // Add additional context for agents
    if (status === 'compilation_started' || status === 'compilation_in_progress') {
      content += `\n\n---\n**For AI Assistants**: This tool will monitor compilation progress. ` +
                `Wait for a completion message before executing other Unity-related commands. ` +
                `If compilation fails, help the user identify and fix the errors shown in the Unity Console.`;
    }

    return {
      content: [
        {
          type: 'text',
          text: content
        }
      ],
      isError: !success
    };
  } catch (error) {
    if (error instanceof McpUnityError) {
      throw error;
    }
      throw new McpUnityError(
      ErrorType.TOOL_EXECUTION,
      `Failed to request script compilation: ${error instanceof Error ? error.message : String(error)}`,
      { originalError: error }
    );
  }
}
